File: Sorting/HeapSort/minHeapSort.cpp
#include<iostream>
using namespace std;

void printArr(int arr[],int n)
{
    for(int i=0;i<n;i++)
    {
        cout<<arr[i]<<"\t";
    }
    cout<<endl;
}

void heapify(int arr[], int n, int i)
{
    int lowest = i;
    int left = 2*i+1;
    int right = 2*i+2;
    if(left<n && arr[left]<arr[lowest])
    {
        lowest=left;
    }
    if(right<n && arr[right]<arr[lowest])
    {
        lowest=right;
    }
    if(lowest!=i)
    {
        swap(arr[lowest],arr[i]);
        heapify(arr,n,lowest);
    }
}

void minHeapSort(int arr[],int n)
{
    for(int i=n/2-1;i>=0;i--)
    {
        heapify(arr,n,i);
        printArr(arr,n);
    }
    cout<<"\nSorted and Swapping\n";
    for(int i=n-1;i>=0;i--)
    {
        swap(arr[0],arr[i]);
        printArr(arr,n);
        heapify(arr,i,0);
    }
}

int main()
{
    int arr[] = {1,12,43,53,1,435,432,4};
    int n = sizeof(arr)/sizeof(arr[0]);
    cout<<"\nInitial Array\n";
    printArr(arr,n);
    cout<<"\nSorting Started\n";
    minHeapSort(arr,n);
    cout<<"\n\n\nfinally sorted array\n\n";
    printArr(arr,n);
}

File: Sorting/HeapSort/minHeapSort.py
def printArr(arr):
    for i in range(len(arr)):
        print(arr[i],end = "\t")
        
    print("\n")
    
def heapify(arr,n,i):
    lowest = i
    left = 2*i + 1 
    right = 2*i + 2
    if(left<n and arr[left]<arr[lowest]):
        lowest=left
    if(right<n and arr[right]<arr[lowest]):
        lowest=right
    if(lowest != i):
        arr[lowest],arr[i] = arr[i],arr[lowest]
        heapify(arr,n,lowest)
        
def heapSort(arr):
    n = len(arr)
    for i in range(n//2,-1,-1):
        heapify(arr,n,i)
        printArr(arr)
        
    print("\nSorted and Swapping")
    
    for i in range(n-1,0,-1):
        arr[i],arr[0] = arr[0],arr[i]
        printArr(arr)
        heapify(arr,i,0)

if __name__ == "__main__":
    data = [42,45,325,643,435,3,576,76,54,34]
    print("\nInitial Arr")
    printArr(data)
    print("\nSorting Started")
    heapSort(data)
    print("\nFinial Data")
    printArr(data)